src/actions/transferActions.js

```diff
@@ -572,6 +573,301 @@ async function _getTransfer (transferId: ?string, receivingId: ?string) {
+    }
+  }
+
+  // tx always have property hash regardless of cryptoType
+
+  // limit the numnber of txs to MAX_TXS
+  txs = txs.slice(0, MAX_TXS)
+
+  // create a dict with key of txHash for existing tx history
+  let historyDict = {}
+  if (accountTxHistory.history) {
+    accountTxHistory.history.map(tx => (historyDict[tx.txHash] = tx))
+  }
+
+  // filter out tx whose corresponding transferData or standaloneTx already
+  // exists in accountTxHistory
+  txs = txs.filter(tx => !historyDict[tx.txHash])
+
+  const txHash2Idx = txs.reduce((accumulator, currentValue, idx) => {
+    if (currentValue && currentValue.hash) {
+      accumulator[currentValue.hash] = idx
+    }
+    return accumulator
+  }, {})
+
+  // lookup hashes in our transfer database to see if any
+  // transfer.[sendTxHash | receiveTxHash | cancelTxHash] matches txHash
+  const txHashLookupResults: Array<{
+    txHash: string,
+    transferId: string,
+    receivingId: string
+  }> = await API.lookupTxHash(txs.map(tx => tx.hash))
+
+  // split results into matched and unmatched
+  const matcehd = {
+    transferIdList: txHashLookupResults
+      .filter(result => result.transferId)
+      .map(result => result.transferId),
+    receivingIdList: txHashLookupResults
+      .filter(result => result.receivingId)
+      .map(result => result.receivingId)
+  }
+  const unmatched = txHashLookupResults.filter(result => !result.transferId && !result.receivingId)
+
+  // gather transferData using transferIds and receivingIds
+  const transferDataList = await API.getBatchTransfers({
+    transferIds: matcehd.transferIdList,
+    receivingIds: matcehd.receivingIdList
+  })
+
+  // handle matched txHash list
+  // fill in txs with transferData
+  for (let transferData of transferDataList) {
+    let idx
+    let { sendTxHash, receiveTxHash, cancelTxHash } = transferData
+    let transferType: string
+    let timestamp
+    // exactly one of [sendTxHash | receiveTxHash | cancelTxHash] matches txHash
+    if (sendTxHash && sendTxHash in txHash2Idx) {
+      idx = txHash2Idx[sendTxHash]
+      transferType = 'SENDER'
+      timestamp = transferData.senderToChainsfer.txTimestamp
+    } else if (receiveTxHash && receiveTxHash in txHash2Idx) {
+      idx = txHash2Idx[receiveTxHash]
+      transferType = 'RECEIVER'
+      timestamp = transferData.chainsferToReceiver.txTimestamp
+    } else if (cancelTxHash && cancelTxHash in txHash2Idx) {
+      idx = txHash2Idx[cancelTxHash]
+      transferType = 'SENDER'
+      timestamp = transferData.chainsferToSender.txTimestamp
+    } else {
+      console.log(sendTxHash, txHash2Idx[sendTxHash])
+      // this should not happen
+      throw new Error('transferData does not have matched txHash')
+    }
+
+    // fill in data
+    transferData = unmarshalAccount(transferData)
+    rv[txs[idx].hash] = { timestamp }
+    rv[txs[idx].hash].transferData = { ...transferData, transferType }
+  }
```

Comment:
put 718 - 746 into a separate function.

Reply:
cannot move due to multiple dependencies

Reply:
cannot move due to multiple dependencies